test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java

```diff
@@ -0,0 +1,1226 @@
+            VERIFY_IR = onlyWhitelistedJTregVMAndJavaOptsFlags();
+            if (!VERIFY_IR) {
+                System.out.println("IR verification disabled due to using non-whitelisted JTreg VM or Javaoptions flag(s).\n");
+            }
+        }
+    }
+
+    /**
+     * For scenarios: Run the tests with the scenario settings and collect all exceptions to be able to run all
+     * scenarios without prematurely throwing an exception. Format violations, however, are wrong for all scenarios
+     * and thus is reported immediately on the first scenario execution.
+     */
+    private void startWithScenarios() {
+        Map<Scenario, Exception> exceptionMap = new TreeMap<>(Comparator.comparingInt(Scenario::getIndex));
+        Set<Integer> scenarioIndices = new HashSet<>();
+        for (Scenario scenario : scenarios) {
+            int scenarioIndex = scenario.getIndex();
+            TestFormat.check(!scenarioIndices.contains(scenarioIndex),
+                             "Cannot define two scenarios with the same index " + scenarioIndex);
+            scenarioIndices.add(scenarioIndex);
+            try {
+                start(scenario);
+            } catch (TestFormatException e) {
+                // Test format violation is wrong for all the scenarios. Only report once.
+                throw e;
+            } catch (Exception e) {
+                exceptionMap.put(scenario, e);
+            }
+        }
+        if (!exceptionMap.isEmpty()) {
+            reportScenarioFailures(exceptionMap);
+        }
+    }
+
+    private void reportScenarioFailures(Map<Scenario, Exception> exceptionMap) {
+        String failedScenarios = "The following scenarios have failed: #"
+                                 + exceptionMap.keySet().stream()
+                                               .map(s -> String.valueOf(s.getIndex()))
+                                               .collect(Collectors.joining(", #"));
+        StringBuilder builder = new StringBuilder(failedScenarios);
+        builder.append("\n\n");
+        for (Map.Entry<Scenario, Exception> entry : exceptionMap.entrySet()) {
+            Exception e = entry.getValue();
+            Scenario scenario = entry.getKey();
+            String errorMsg = "";
+            if (scenario != null) {
+                errorMsg = getScenarioTitleAndFlags(scenario);
+            }
+            if (e instanceof IRViolationException) {
+                IRViolationException irException = (IRViolationException) e;
+                // For IR violations, only show the actual violations and not the (uninteresting) stack trace.
+                System.out.println((scenario != null ? "Scenario #" + scenario.getIndex() + " - " : "")
+                                   + "Compilation(s) of failed matche(s):");
+                System.out.println(irException.getCompilations());
+                builder.append(errorMsg).append("\n").append(irException.getExceptionInfo()).append(e.getMessage());
```

Comment:
as `builder.toString` will be printed out to cout/cerr, you should use platform-specific line-separator instead of `\n`